### PR TITLE
feat(#2349): add "right_align" option for renderer.icons.*_placement

### LINKED ISSUE #2349: Icon Placement Right
**Can this functionality be implemented utilising API?**
No, it's internal behaviour.

**Describe the solution you'd like**
Show icons: git, diagnostic, modified at the far right side of the tree.

**Describe alternatives you've considered**
Right sign column, not possible.

**Additional context**
#2348

### PR DESCRIPTION
**Pull request by evertonse**
Closes #2349

#### This PR augments options for icons placement allowing rendering of theses decorations on the far right using the ``right_align`` option from neovim ``ext_marks``

image

##  Setup for testing
```lua
require('nvim-tree').setup {
  renderer = {
    icons = {
      git_placement = 'right_align',
      modified_placement = 'right_align',
      diagnostics_placement = 'right_align',
      bookmarks_placement = 'right_align',
      padding = ' ',
      show = {
        file = true,
        folder = true,
        diagnostics = true,
        bookmarks = true,
        git = true,
        modified = true,
      },
    },
  },
}
```

If there's anything you'd like me to change about documentation or code, let me know, I'll do it.

### CODE REVIEW

**Review by alex-courtis — COMMENTED** [review]
This looks incredible - I'll be using this one right away.

One question before I merge: What's the significance of the
```
`"right_align"` (requires |nvim_buf_set_extmark|).
```
line in the help?

Is there something that needs to be enabled / present? I tested on nvim 0.9.5 and it worked perfectly.

**Comment by evertonse** [comment]
> nvim_buf_set_extmark

That's jsut to refer that the version of neovim has to have ``nvim_buf_set_extmark`` in the api, because that Idk when was ``extmarks`` introduced in neovim.
Neovim >= 0.9 must be all good.

**Comment by alex-courtis** [comment]
> > nvim_buf_set_extmark
> 
> That's just to refer that the version of neovim has to have `nvim_buf_set_extmark` in the api, because that Idk when was `extmarks` introduced in neovim. Neovim >= 0.9 must be all good.

All good; we recently updated to >= 0.9

**Review by alex-courtis — APPROVED** [review]
In the interest of time I consolidated the doc and removed the `nvim_buf_set_extmark` note.

Many thanks for your work @evertonse - I'll enjoy using this one.

**Comment by alex-courtis** [comment]
Deployed and hit a problem with full name, when there are right icons and the name is truncated.

Looking at a quick fix...
```
Error detected while processing CursorMoved Autocommands for "NvimTree_*":
Error executing lua callback: ...tree.lua/lua/nvim-tree/renderer/components/full-name.lua:71: Error executing lua: ...tree.lua/lua/nvim-tree/renderer/components/full-name.lua:82: Invalid 'col_end': Expected Lua number
stack traceback:
        [C]: in function 'nvim_buf_add_highlight'
        ...tree.lua/lua/nvim-tree/renderer/components/full-name.lua:82: in function <...tree.lua/lua/nvim-tree/renderer/components/full-name.lua:71>
        [C]: in function 'nvim_win_call'
        ...tree.lua/lua/nvim-tree/renderer/components/full-name.lua:71: in function 'show'
        ...tree.lua/lua/nvim-tree/renderer/components/full-name.lua:110: in function <...tree.lua/lua/nvim-tree/renderer/components/full-name.lua:108>
stack traceback:
        [C]: in function 'nvim_win_call'
        ...tree.lua/lua/nvim-tree/renderer/components/full-name.lua:71: in function 'show'
        ...tree.lua/lua/nvim-tree/renderer/components/full-name.lua:110: in function <...tree.lua/lua/nvim-tree/renderer/components/full-name.lua:108>
```

**Comment by alex-courtis** [comment]
It looks like we have a collision with the extmarks used by full-name:

https://github.com/evertonse/nvim-tree.lua/blob/26632f496e7e3c0450d8ecff88f49068cecc8bda/lua/nvim-tree/renderer/components/full-name.lua#L70

Sorry @evertonse I'm going to have to revert this one; I can't see a quick/safe fix.

Can I please leave it to you?

Minimal config:
```lua
require("nvim-tree").setup({
  renderer = {
    full_name = true,
    icons = {
      git_placement = "right_align",
    },
  },
})
```

nvim-tree.lua modified, `j` on the folder above it

**Comment by evertonse** [comment]
I was able to reproduce it, took me while to understand this 😅 MB
> nvim-tree.lua modified, `j` on the folder above it

I'll fix it someday this week. Should I make a separate PR?

**Pull request merged**
